### DISCUSSION TITLE
Testing equality of function with two return values

### DISCUSSION THREAD

**Discussion by pwaring** [stretchr/testify#1683] [Q&A] [answered]
I have the following function which returns two values:

```go
func getUnsortedPairs(input string) (left []int, right []int) {
	left = []int{}
	right = []int{}
	pairs := strings.Split(input, "\n")

	for p := range pairs {
		if len(pairs[p]) > 0 {
			parts := strings.Fields(pairs[p])
			leftPart, _ := strconv.Atoi(parts[0])
			rightPart, _ := strconv.Atoi(parts[1])

			left = append(left, leftPart)
			right = append(right, rightPart)
		}
	}

	return left, right
}
```

I want to check that the two returned values match two expected values, ideally in a single assertion (I can split them across two assertions but then they're not tightly coupled, and they should be tested as a single 'what does this function return' check).

This doesn't work, as it makes Equals think that the two return types are separate parameters:

```go
leftExpected := []int{}
rightExpected := []int{}
assert.Equal(t, leftExpected, rightExpected, getUnsortedPairs(""))
```

This also doesn't work, with a syntax error:

```go
leftExpected := []int{}
rightExpected := []int{}
assert.Equal(t, (leftExpected, rightExpected), getUnsortedPairs(""))
```

Is there a way to do what I want with a single assertion, or do I have to write my own assertion function or call assert.Equals for each return value?

**Comment by brackendawson** [comment, marked as answer]
The idiomatic way to implement this in Go is this:
```go
leftExpected := []int{}
rightExpected := []int{}
left, right := getUnsortedPairs("")
assert.Equal(t, leftExpected, left)
assert.Equal(t, rightExpected, right)
```
This is tightly coupled.

**Reply by brackendawson** [reply]
You can pretend Go has tuples and try to write a single assertion if you want, but it is absolutely disgusting.
```go
	expected := [2][]int{{}, {}}
	assert.Equal(t, expected, func() (tpl [2][]int) { tpl[0], tpl[1] = getUnsortedPairs(""); return }())
```
